pkg/endpoint/local_ip.go

```diff
@@ -53,12 +53,25 @@ func getLocalIPFromRoutes(family k8snet.IPFamily) (string, error) {
 }
 
 func GetLocalIPForDestination(dst string, family k8snet.IPFamily) string {
-	conn, err := Dial("udp"+string(family), "["+dst+"]:53")
+	conn, err := Dial("udp"+string(family), net.JoinHostPort(dst, "53"))
 	if err == nil {
 		defer conn.Close()
 		localAddr := conn.LocalAddr().(*net.UDPAddr)
 
-		return localAddr.IP.String()
+		if family == k8snet.IPv4 || isValidGlobalIPv6(localAddr.IP) {
+			return localAddr.IP.String()
+		}
+
+		logger.Warningf("IP %q returned from Dial isn't usable - trying to find another IP from the same interface", localAddr.IP)
+
+		// Try to find a valid global-scope IP from the same interface
```

Comment:
In case we failed to find validIPv6Adresses , we can fallback to getLocalIPFromRoutes, and we can remove getValidGlobalIPv6FromSameInterface , getLocalIPFromRoutes function should be something like : 

```
unc getLocalIPFromRoutes(family k8snet.IPFamily) (string, error) {
	netlink := netlinkAPI.New()

	routes, err := netlink.RouteList(nil, family)
	if err != nil {
		return "", errors.Wrapf(err, "error listing routes for IPv%v", family)
	}

	for i := range routes {
		if routes[i].Gw != nil && routes[i].Src != nil {
			ipStr := routes[i].Src.String()

			if  family == k8snet.IPv6  && !isValidGlobalIPv6(ipStr) {
				continue
			}

			return ipStr, nil
		}
	}

	return "", errors.Errorf("couldn't find route for family %v", family)
}
```

Reply:
Yes, that makes it robust. Did the changes.